src/ErrorHandler.php

```diff
@@ -1,120 +1,93 @@
 <?php
-declare(strict_types = 1);
+declare(strict_types=1);
 
 namespace Middlewares;
 
+use Middlewares\Formatter\FormatterInterface;
+use Middlewares\Formatter\PlainFormatter;
+use Middlewares\Utils\Factory;
+use Psr\Http\Message\ResponseFactoryInterface;
 use Psr\Http\Message\ResponseInterface;
 use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Message\StreamFactoryInterface;
 use Psr\Http\Server\MiddlewareInterface;
 use Psr\Http\Server\RequestHandlerInterface;
 use Throwable;
 
 class ErrorHandler implements MiddlewareInterface
 {
-    private $handler;
+    /** @var ResponseFactoryInterface */
+    private $responseFactory;
 
-    /**
-     * @var callable|null The status code validator
-     */
-    private $statusCodeValidator;
-
-    /**
-     * @var bool Whether or not catch exceptions
-     */
-    private $catchExceptions = false;
-
-    /**
-     * @var string The attribute name
-     */
-    private $attribute = 'error';
-
-    public function __construct(RequestHandlerInterface $handler = null)
-    {
-        $this->handler = $handler;
-    }
-
-    /**
-     * Configure the catchExceptions.
-     */
-    public function catchExceptions(bool $catch = true): self
-    {
-        $this->catchExceptions = (bool) $catch;
+    /** @var StreamFactoryInterface */
+    private $streamFactory;
 
-        return $this;
-    }
-
-    /**
-     * Configure the status code validator.
-     */
-    public function statusCode(callable $statusCodeValidator): self
-    {
-        $this->statusCodeValidator = $statusCodeValidator;
+    /** @var FormatterInterface[] */
+    private $formatters = [];
 
-        return $this;
+    public function __construct(
+        ResponseFactoryInterface $responseFactory = null,
+        StreamFactoryInterface $streamFactory = null
+    ) {
+        $this->responseFactory = $responseFactory ?? Factory::getResponseFactory();
+        $this->streamFactory = $streamFactory ?? Factory::getStreamFactory();
     }
```

Comment:
I prefer to use methods to configure the factories instead inject them in the constructor, for api consistency with the rest of middlewares. There are Traits in middlewares/utils for this purpose: https://github.com/middlewares/utils/tree/master/src/Traits.

Reply:
This topic was discused here: https://github.com/middlewares/fast-route/pull/9#issuecomment-409985531. The factories are not always needed, they are optional. There are many middlewares requiring factories, some for trivial purposes (for example, create redirect responses only [under specific conditions](https://github.com/middlewares/trailing-slash/blob/master/src/TrailingSlash.php#L53) and others that make a more expensive use of factories (like [read and serve static files](https://github.com/middlewares/filesystem/blob/master/src/Reader.php)).
I try to maintain all these middlewares from more global point of view, in which they share some kind of consistency in the API.

Reply:
Easier doesn't make it right. Not knowing the dependencies makes it harder to use, IMO. Especially if you use auto-wiring as a dependency injection strategy.

For me this is kind of a deal breaker. I really don't want to dilute my code and make it harder to use with auto-wiring. I totally understand where you are coming from with the trait, but all it does is shift the burden of managing dependencies from you to everyone else. That's a 👎 in my book.

I would ask you to reconsider this stance or I am likely going to put my efforts in a different direction.